File: src/templates/backend/fastapi/src/api/user_api.py
from fastapi import APIRouter, HTTPException, status
from src.service.user_service import user_service
from src.types.user_type import CreateUserDto, UpdateUserDto, ApiResponse
from src.function.helper import validate_create_user_dto, validate_update_user_dto, validate_string

router = APIRouter(prefix="/users", tags=["Users"])


@router.get("/", response_model=ApiResponse)
async def get_all_users():
    try:
        users = await user_service.get_all_users()
        return ApiResponse(
            success=True,
            data=users,
            message="Users retrieved successfully",
        )
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/{user_id}", response_model=ApiResponse)
async def get_user_by_id(user_id: int):
    try:
        if user_id <= 0:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid user ID")

        user = await user_service.get_user_by_id(user_id)
        if not user:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")

        return ApiResponse(
            success=True,
            data=user,
            message="User retrieved successfully",
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.get("/search/{keyword}", response_model=ApiResponse)
async def search_users(keyword: str):
    try:
        if not validate_string(keyword, 1, 100):
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Search keyword must be between 1 and 100 characters")

        users = await user_service.search_users(keyword)
        return ApiResponse(
            success=True,
            data=users,
            message="Users retrieved successfully",
        )
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.post("/", response_model=ApiResponse, status_code=status.HTTP_201_CREATED)
async def create_user(data: dict):
    try:
        validated_data = validate_create_user_dto(data)
        user = await user_service.create_user(validated_data)
        return ApiResponse(
            success=True,
            data=user,
            message="User created successfully",
        )
    except ValueError as e:
        error_message = str(e)
        if "already exists" in error_message or "already taken" in error_message:
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=error_message)
        else:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=error_message)
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.put("/{user_id}", response_model=ApiResponse)
async def update_user(user_id: int, data: dict):
    try:
        if user_id <= 0:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid user ID")

        validated_data = validate_update_user_dto(data)
        user = await user_service.update_user(user_id, validated_data)
        if not user:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")

        return ApiResponse(
            success=True,
            data=user,
            message="User updated successfully",
        )
    except ValueError as e:
        error_message = str(e)
        if "already in use" in error_message or "already taken" in error_message:
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=error_message)
        elif "not found" in error_message:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=error_message)
        else:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=error_message)
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.delete("/{user_id}", response_model=ApiResponse)
async def delete_user(user_id: int):
    try:
        if user_id <= 0:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid user ID")

        deleted = await user_service.delete_user(user_id)
        if not deleted:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")

        return ApiResponse(
            success=True,
            message="User deleted successfully",
        )
    except ValueError as e:
        error_message = str(e)
        if "not found" in error_message:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=error_message)
        else:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=error_message)
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))


@router.post("/{user_id}/soft-delete", response_model=ApiResponse)
async def soft_delete_user(user_id: int):
    try:
        if user_id <= 0:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid user ID")

        user = await user_service.soft_delete_user(user_id)
        if not user:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")

        return ApiResponse(
            success=True,
            data=user,
            message="User soft deleted successfully",
        )
    except ValueError as e:
        error_message = str(e)
        if "not found" in error_message:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=error_message)
        else:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=error_message)
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e))

File: src/templates/backend/elysia/src/api/user_api.ts
import { Elysia, t } from "elysia";
import { userService } from "../service/user_service";
import { validateString, validateCreateUserDto, validateUpdateUserDto } from "../function/helper";
import type { ApiResponse, UserResponse } from "../types/user_type";

export const userApi = new Elysia({ prefix: "/users" })
  .get(
    "/",
    async () => {
      try {
        const users = await userService.getAllUsers();
        return {
          success: true,
          data: users,
          message: "Users retrieved successfully",
        } as ApiResponse<UserResponse[]>;
      } catch (error) {
        throw {
          success: false,
          error: error instanceof Error ? error.message : String(error),
        } as ApiResponse<never>;
      }
    },
    {
      detail: {
        tags: ["Users"],
        summary: "Get all users",
        description: "Retrieve all active users from the database",
      },
    }
  )

  .get(
    "/:id",
    async ({ params, set }) => {
      try {
        const id = parseInt(params.id);
        if (isNaN(id) || id <= 0) {
          set.status = 400;
          throw new Error("Invalid user ID");
        }

        const user = await userService.getUserById(id);
        if (!user) {
          set.status = 404;
          throw new Error("User not found");
        }

        return {
          success: true,
          data: user,
          message: "User retrieved successfully",
        } as ApiResponse<UserResponse>;
      } catch (error) {
        throw {
          success: false,
          error: error instanceof Error ? error.message : String(error),
        } as ApiResponse<never>;
      }
    },
    {
      params: t.Object({
        id: t.String(),
      }),
      detail: {
        tags: ["Users"],
        summary: "Get user by ID",
        description: "Retrieve a specific user by their ID",
      },
    }
  )

  .get(
    "/search/:keyword",
    async ({ params }) => {
      try {
        const keyword = params.keyword;
        if (!validateString(keyword, 1, 100)) {
          throw new Error("Search keyword must be between 1 and 100 characters");
        }

        const users = await userService.searchUsers(keyword);
        return {
          success: true,
          data: users,
          message: "Users retrieved successfully",
        } as ApiResponse<UserResponse[]>;
      } catch (error) {
        throw {
          success: false,
          error: error instanceof Error ? error.message : String(error),
        } as ApiResponse<never>;
      }
    },
    {
      params: t.Object({
        keyword: t.String(),
      }),
      detail: {
        tags: ["Users"],
        summary: "Search users",
        description: "Search for users by email, username, or full name",
      },
    }
  )

  .post(
    "/",
    async ({ body, set }) => {
      try {
        const validatedData = validateCreateUserDto(body);
        const user = await userService.createUser(validatedData);
        set.status = 201;
        return {
          success: true,
          data: user,
          message: "User created successfully",
        } as ApiResponse<UserResponse>;
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        if (errorMessage.includes("already exists") || errorMessage.includes("already taken")) {
          set.status = 409;
        } else if (errorMessage.includes("Validation error")) {
          set.status = 400;
        } else {
          set.status = 400;
        }
        throw {
          success: false,
          error: errorMessage,
        } as ApiResponse<never>;
      }
    },
    {
      body: t.Object({
        email: t.String(),
        username: t.String(),
        full_name: t.String(),
        password: t.String(),
        avatar_url: t.Optional(t.String()),
        bio: t.Optional(t.String()),
        role: t.Optional(t.String()),
      }),
      detail: {
        tags: ["Users"],
        summary: "Create a new user",
        description: "Create a new user with the provided data",
      },
    }
  )

  .put(
    "/:id",
    async ({ params, body, set }) => {
      try {
        const id = parseInt(params.id);
        if (isNaN(id) || id <= 0) {
          set.status = 400;
          throw new Error("Invalid user ID");
        }

        const validatedData = validateUpdateUserDto(body);
        const user = await userService.updateUser(id, validatedData);
        if (!user) {
          set.status = 404;
          throw new Error("User not found");
        }

        return {
          success: true,
          data: user,
          message: "User updated successfully",
        } as ApiResponse<UserResponse>;
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        if (errorMessage.includes("already in use") || errorMessage.includes("already taken")) {
          set.status = 409;
        } else if (errorMessage.includes("not found")) {
          set.status = 404;
        } else if (errorMessage.includes("Validation error")) {
          set.status = 400;
        } else {
          set.status = 400;
        }
        throw {
          success: false,
          error: errorMessage,
        } as ApiResponse<never>;
      }
    },
    {
      params: t.Object({
        id: t.String(),
      }),
      body: t.Object({
        email: t.Optional(t.String()),
        username: t.Optional(t.String()),
        full_name: t.Optional(t.String()),
        password: t.Optional(t.String()),
        avatar_url: t.Optional(t.String()),
        bio: t.Optional(t.String()),
        role: t.Optional(t.String()),
        is_active: t.Optional(t.Boolean()),
      }),
      detail: {
        tags: ["Users"],
        summary: "Update a user",
        description: "Update an existing user's information",
      },
    }
  )

  .delete(
    "/:id",
    async ({ params, set }) => {
      try {
        const id = parseInt(params.id);
        if (isNaN(id) || id <= 0) {
          set.status = 400;
          throw new Error("Invalid user ID");
        }

        const deleted = await userService.deleteUser(id);
        if (!deleted) {
          set.status = 404;
          throw new Error("User not found");
        }

        return {
          success: true,
          message: "User deleted successfully",
        } as ApiResponse<never>;
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        if (errorMessage.includes("not found")) {
          set.status = 404;
        } else {
          set.status = 400;
        }
        throw {
          success: false,
          error: errorMessage,
        } as ApiResponse<never>;
      }
    },
    {
      params: t.Object({
        id: t.String(),
      }),
      detail: {
        tags: ["Users"],
        summary: "Delete a user",
        description: "Permanently delete a user from the database",
      },
    }
  )

  .post(
    "/:id/soft-delete",
    async ({ params, set }) => {
      try {
        const id = parseInt(params.id);
        if (isNaN(id) || id <= 0) {
          set.status = 400;
          throw new Error("Invalid user ID");
        }

        const user = await userService.softDeleteUser(id);
        if (!user) {
          set.status = 404;
          throw new Error("User not found");
        }

        return {
          success: true,
          data: user,
          message: "User soft deleted successfully",
        } as ApiResponse<UserResponse>;
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : String(error);
        if (errorMessage.includes("not found")) {
          set.status = 404;
        } else {
          set.status = 400;
        }
        throw {
          success: false,
          error: errorMessage,
        } as ApiResponse<never>;
      }
    },
    {
      params: t.Object({
        id: t.String(),
      }),
      detail: {
        tags: ["Users"],
        summary: "Soft delete a user",
        description: "Deactivate a user without permanently deleting them",
      },
    }
  )

  .onError(({ error, set }) => {
    if (error && typeof error === "object" && "success" in error) {
      return error;
    }
    set.status = 500;
    return {
      success: false,
      error: error instanceof Error ? error.message : "Internal server error",
    } as ApiResponse<never>;
  });
